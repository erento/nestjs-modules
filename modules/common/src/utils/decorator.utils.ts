/* tslint:disable:variable-name */
import {applyDecorators, Header} from '@nestjs/common';
type DecoratorFunction = (...args: any[]) => MethodDecorator;

export const PrivateCache: DecoratorFunction = (): MethodDecorator => applyDecorators(
    Header('Cache-Control', 'private'),
    Header('Surrogate-Control', 'private'),
);

export const CacheForSeconds: DecoratorFunction = (duration: number): MethodDecorator => applyDecorators(
    Header('Cache-Control', `max-age=${duration}`),
);
