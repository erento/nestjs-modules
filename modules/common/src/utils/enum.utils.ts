const getCleanStringKeys: Function = (enumeration: any): string[] => {
    const objKeys: string[] = Object.keys(enumeration);

    return objKeys
        .filter((value: string): boolean => isNaN(+value))
        .map((value: string): string => enumeration[value]);
};

export const getEnumValues: (enumeration: any) => string[] = (enumeration: any): string[] => {
    const enumerationObj: string[] = JSON.parse(JSON.stringify(enumeration));

    return getCleanStringKeys(enumerationObj);
};

export const getMaxLengthOfEnumValues: (enumeration: any) => number = (enumeration: any): number => {
    const enumerationValuesLengths: number[] = getEnumValues(enumeration)
        .map((enumerationValue: string): number => enumerationValue.length);

    if (enumerationValuesLengths.indexOf(<any> undefined) !== -1) {
        throw new Error(`Enumeration can contain only string values to get a max length.`);
    }

    return Math.max(...enumerationValuesLengths);
};
