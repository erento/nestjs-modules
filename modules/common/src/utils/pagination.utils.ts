import {OffsetAndLimit, PaginatedResponse, PaginationParams} from '../pipes/interfaces';

export function getPaginationQuery (params?: PaginationParams): OffsetAndLimit {
    if (!params) {
        return {};
    }

    return {
        offset: Math.max(params.page - 1, 0) * params.size,
        limit: params.size,
    };
}

export function mapToPaginationResponse<T> (paginationParams: PaginationParams, count: number, results: T[]): PaginatedResponse<T> {
    return {
        page: paginationParams.page,
        pages: Math.ceil(count / paginationParams.size),
        results,
        size: paginationParams.size,
        total: count,
    };
}
