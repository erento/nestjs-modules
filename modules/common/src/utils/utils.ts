import {Environments} from '../environments/environments';
import {UtilsError} from './errors/utils.error';

interface ES7Array extends Array<any> {
    includes (value: any): boolean;
}

export function createBuffer (body: string, encoding: BufferEncoding = 'base64'): Buffer {
    try {
        return Buffer.from(body, encoding);
    } catch {
        throw new UtilsError('Cannot create a buffer.');
    }
}

export function deepCopyObj<T> (sourceObj: T): T | undefined {
    try {
        return JSON.parse(JSON.stringify(sourceObj));
    } catch {
        return undefined;
    }
}

export function arrayIntersection (arr1: any[], arr2: any[]): any[] {
    return arr1.filter((x: any): boolean => (<ES7Array> arr2).includes(x));
}

export function arrayDifference (arr1: any[], arr2: any[]): any[] {
    return arr1
        .filter((x: any): boolean => !(<ES7Array> arr2).includes(x))
        .concat(arr2.filter((x: any): boolean => !(<ES7Array> arr1).includes(x)));
}

export function getPropertyOrUndefined<T> (property: T | null | undefined): T | undefined {
    return (property === null || property === undefined) ? undefined : property;
}

export function getTrueOrUndefined (value: boolean): true | undefined {
    return !value ? undefined : value;
}

export function getDateStringOrUndefined (date: string | Date | null): string | undefined {
    if (date === null) {
        return undefined;
    }

    const parsedDate: Date = new Date(date);
    if (isNaN(parsedDate.valueOf()) || parsedDate.getTime() === 0) {
        return undefined;
    }

    return parsedDate.toISOString();
}

export function parseDateOrUndefined (date?: string): Date | undefined {
    if (date === undefined) {
        return undefined;
    }

    try {
        const parsed: number = Date.parse(date);
        if (isNaN(parsed)) {
            return undefined;
        }
        return new Date(parsed);
    } catch (e) {
        return undefined;
    }
}

export function isUuid (value: string | null | undefined): boolean {
    return !!value && /^[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}$/.test(value);
}

// is used for fields which are saved as JSONB. NULL values have to be cleaned on reading
// default has to be undefined, so the property is not returned
export function cleanNullValues<T extends Object> (someObject: T | null | undefined): T | undefined {
    if (someObject === null || someObject === undefined) {
        return undefined;
    }

    for (const propName in someObject) {
        if (someObject.hasOwnProperty(propName) && <any> someObject[propName] === null) {
            delete someObject[propName];
        }
    }

    if (Object.keys(someObject).length === 0) {
        return undefined;
    }

    return someObject;
}

// Is used for apis which provide empty string values as default values.
// Those defaults are converted to null to be able to remove these fields in the database
export function replaceEmptyStringValuesWithNull<T extends Object> (someObject: T | null): T | null {
    if (someObject === null || Object.keys(someObject).length === 0) {
        return null;
    }

    Object.keys(someObject).forEach((key: string): void => {
        if (!someObject[key] || (undefined !== someObject[key].trim && someObject[key].trim() === '')) {
            someObject[key] = null;
        }
    });

    return someObject;
}

// This is for nullable values, false is excluded because it is a valid value
export function getPropertyOrFallbackOrUndefined<V, F> (value: V | undefined | null, fallBack?: F | null): V | F | undefined {
    if (value !== undefined && value !== null && <unknown> value !== '') {
        return value;
    }

    if (fallBack !== undefined && <unknown> fallBack !== '' && fallBack !== null) {
        return fallBack;
    }

    return undefined;
}

// This is for not-nullable values, false is excluded because it is a valid value
export function getPropertyOrFallback<V, F> (value: V | undefined | null, fallBack: F): V | F {
    if (value !== undefined && value !== null && <unknown> value !== '') {
        return value;
    }

    return fallBack;
}

export function fillArrayWithIndices (length: number = 0): number[] {
    return [...Array(length).keys()];
}

export function batchArray<T> (items: T[], maxBatchSize: number): T[][] {
    const batches: T[][] = [];

    let currentBatch: number = 0;
    items.forEach((item: T): void => {
        if (batches[currentBatch] && batches[currentBatch].length >= maxBatchSize) {
            ++currentBatch;
        }
        if (!batches[currentBatch]) {
            batches[currentBatch] = [];
        }
        batches[currentBatch].push(item);
    });

    return batches;
}

export function requiredEnvVariable (key: keyof typeof process.env): string {
    if (key in process.env && process.env[key]) {
        return <string> process.env[key];
    }

    if (Environments.isTest()) {
        // enables running test directly in IDE in test mode
        return '';
    }

    throw new Error(`Missing environment variable "${key}"!`);
}

export function oneLine (multiline: string): string {
    return multiline
        .replace(/(?:\n(?:\s*))+/g, ' ')
        .trim();
}
