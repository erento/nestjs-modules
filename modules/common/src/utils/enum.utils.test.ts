import {getEnumValues, getMaxLengthOfEnumValues} from './enum.utils';

enum EnumEmpty {}

enum EnumWithStrings {
    A = 'some text',
    B = 'some longer text',
    C = 'some even longer text',
}

enum EnumWithNumbers {
    A = 1,
    B = 100,
    C = 10000,
}

enum EnumWithMixedValues {
    A = 1,
    B = 100,
    C = 10000,
    D = 'some text',
    E = 'some longer text',
    F = 'some even longer text',
}

describe('Enum utils', (): void => {
    test('should get values from empty enumeration', async (): Promise<void> => {
        expect(getEnumValues(EnumEmpty)).toEqual([]);
    });

    test('should get values from string enumeration', async (): Promise<void> => {
        expect(getEnumValues(EnumWithStrings)).toMatchSnapshot();
    });

    test('should get values from number enumeration', async (): Promise<void> => {
        expect(getEnumValues(EnumWithNumbers)).toMatchSnapshot();
    });

    test('should get values from mixed enumeration', async (): Promise<void> => {
        expect(getEnumValues(EnumWithMixedValues)).toMatchSnapshot();
    });

    test('should get max length of enumeration', async (): Promise<void> => {
        expect(getMaxLengthOfEnumValues(EnumWithStrings)).toBe(21);
        expect((): number => getMaxLengthOfEnumValues(EnumWithNumbers)).toThrowErrorMatchingSnapshot();
        expect((): number => getMaxLengthOfEnumValues(EnumWithMixedValues)).toThrowErrorMatchingSnapshot();
    });
});
