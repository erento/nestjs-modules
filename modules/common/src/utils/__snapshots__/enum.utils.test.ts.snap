// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enum utils should get max length of enumeration 1`] = `"Enumeration can contain only string values to get a max length."`;

exports[`Enum utils should get max length of enumeration 2`] = `"Enumeration can contain only string values to get a max length."`;

exports[`Enum utils should get values from mixed enumeration 1`] = `
Array [
  1,
  100,
  10000,
  "some text",
  "some longer text",
  "some even longer text",
]
`;

exports[`Enum utils should get values from number enumeration 1`] = `
Array [
  1,
  100,
  10000,
]
`;

exports[`Enum utils should get values from string enumeration 1`] = `
Array [
  "some text",
  "some longer text",
  "some even longer text",
]
`;
